Mathlib/NumberTheory/LSeries/AbstractFuncEq.lean

```diff
@@ -0,0 +1,450 @@
+  · refine IsBigO.of_bound ‖P.f₀‖ (eventually_nhdsWithin_iff.mpr ?_)
+    filter_upwards [eventually_le_nhds zero_lt_one] with x hx' (hx : 0 < x)
+    apply le_mul_of_one_le_right (norm_nonneg _)
+    rw [norm_of_nonneg (rpow_pos_of_pos hx _).le, rpow_neg hx.le]
+    exact one_le_inv (rpow_pos_of_pos hx _) (rpow_le_one hx.le hx' P.hk.le)
+
+end WeakFEPair
+
+namespace StrongFEPair
+
+variable (P : StrongFEPair E)
+
+/-- As `x → 0`, `F x` decays faster than any power of `x`. -/
+lemma hf_top' (r : ℝ) :
+    P.f =O[atTop] (· ^ (-r)) := by
+  simpa only [P.hf₀, sub_zero] using P.hf_top r
+
+/-- As `x → 0`, `F x` decays faster than any power of `x`. -/
+lemma hf_zero' (r : ℝ) :
+    P.f =O[𝓝[>] 0] (· ^ (-r)) := by
+  have := P.hg₀ ▸ P.hf_zero r
+  simpa only [smul_zero, sub_zero]
+
+/-!
+## Main theorems on strong FE-pairs
+-/
+
+/-- The completed L-function. -/
+def Λ : ℂ → E := mellin P.f
+
+/-- The Mellin transform of `f` is well-defined and equal to `P.Λ s`, for all `s`. -/
+lemma hasMellin (s : ℂ) :
+    HasMellin P.f s (P.Λ s):=
+  let ⟨t, ht⟩ := exists_gt s.re
+  let ⟨u, hu⟩ := exists_lt s.re
+  ⟨mellinConvergent_of_isBigO_rpow P.hf_int (P.hf_top' t) ht (P.hf_zero' u) hu, rfl⟩
+
+/-- If `(f, g)` are a strong FE pair, then the Mellin transform of `f` is entire. -/
+lemma differentiable_Λ : Differentiable ℂ P.Λ := fun s ↦
+  let ⟨_, ht⟩ := exists_gt s.re
+  let ⟨_, hu⟩ := exists_lt s.re
+  mellin_differentiableAt_of_isBigO_rpow P.hf_int (P.hf_top' _) ht (P.hf_zero' _) hu
+
+/-- Main theorem about strong FE pairs: if `(f, g)` are a strong FE pair, then the Mellin
+transforms of `f` and `g` are related by `s ↦ k - s`. -/
+lemma functional_equation (s : ℂ) :
+    P.Λ (P.k - s) = P.ε • P.symm.Λ s := by
+  change mellin P.f (P.k - s) = P.ε • mellin P.g s
+  have step1 := mellin_comp_rpow P.g (-s) (-1)
+  simp_rw [abs_neg, abs_one, inv_one, one_smul, ofReal_neg, ofReal_one, div_neg, div_one, neg_neg,
+    rpow_neg_one, ← one_div] at step1
+  have step2 := mellin_cpow_smul (fun t ↦ P.g (1 / t)) (P.k - s) (-P.k)
+  rw [← sub_eq_add_neg, sub_right_comm, sub_self, zero_sub] at step2
+  simp_rw [step1] at step2
+  have step3 := mellin_const_smul (fun t ↦ (t : ℂ) ^ (-P.k : ℂ) • P.g (1 / t)) (P.k - s) P.ε
+  rw [step2] at step3
+  rw [← step3]
+  refine set_integral_congr measurableSet_Ioi (fun t ht ↦ ?_)
+  simp_rw [P.h_feq' t ht, ← mul_smul]
+  rw [cpow_neg, ofReal_cpow (le_of_lt ht)]
+  have : (t : ℂ) ^ (P.k : ℂ) ≠ 0 := by
+    simpa only [← ofReal_cpow (le_of_lt ht), ofReal_ne_zero] using (rpow_pos_of_pos ht _).ne'
+  field_simp [P.hε]
+
+end StrongFEPair
+
+namespace WeakFEPair
+
+variable (P : WeakFEPair E)
+
+/-!
+## Auxiliary results II: building a strong FE-pair from a weak FE-pair
+-/
+
+/-- Piecewise modified version of `f` with optimal asymptotics. We deliberately choose intervals
+which don't quite join up, so the function is `0` at `x = 1`, in order to maintain symmetry;
+there is no `good` choice of value at `1`. -/
```

Comment:
Then why not defined a parametrised version (say, `f_mod'`) where `1` is a variable, in case we later need to split this at another point? You can then use an `abbreviation` to set this to `1` in the rest of your file. Probably I would also stick to the whole "modified" or "modif" to avoid confusion with "mod" for modular/modulo.

Reply:
There's two separate issues here – choosing where to split the integral, and choosing what value to assign for `f_mod` at the split point. The comment is about the latter, but I think you are perhaps misreading it as the former (which is a much more mathematically interesting issue!).


For the former: I did consider adding an extra parameter `N` in the `FEPair` class, allowing a functional equation relating `t` and `N / t`, and splitting the integral at `sqrt N`; but it all became quite messy. I think it'll be easier just to scale the variables when applying this to higher-level modular forms.

I think the latter (boring) issue, assigning the value at the endpoint, is not likely to be of interest outside the context of this particular proof.

Reply:
If we allow a general "level" parameter `0 < N`, and let the functional equation hypothesis be `f (N / t) = (stuff) * g x`, then what should the "stuff" be? If we have
````
f (N / x) = ε * x ^ k * g x
````
then we lose the symmetry (`P.symm.ε = P.ε⁻¹`) which holds with the current setup. If we make it
````
f (N / x) = ε * (x / sqrt(N)) ^ k * g x
````
then the symmetry is restored, but the proofs of the intermediate results become yet more complicated.

Reply:
How about the following? If we find in future that the corresponding results for non-trivial "level" parameters are useful, we can add (in a future PR) classes `WeakFEPairWithLevel` and `StrongFEPairWithLevel`. Then we can prove some simple lemma to the effect that if `f, g` are a `WeakFEPairWithLevel` N, then `f (· / sqrt N), g (· / sqrt N)` are a `WeakFEPair` of level 1, and level N generalisations of all the results on FE-pairs will follow easily.

Reply:
Sure, that makes sense. I was suggesting a very low-level modifications, namely simply adding the `modif_withLevel` modified version and immediately after setting an abbreviation for `mod_f` (or modif_f) being the `modif_withLevel 1`. But I agree that without adding any further lemma this is not so useful, so your remark is perfectly fine: you could add a `ToDo` line at the beginning with the above remark.

Reply:
I've added a TODO comment.

Reply:
Great, IMHO this PR is ready to go.